cmd/syncthing/gui.go

```diff
@@ -334,29 +338,33 @@ func (s *apiService) Serve() {
 	// Add the CORS handling
 	handler = corsMiddleware(handler, guiCfg.InsecureAllowFrameLoading)
 
-	if addressIsLocalhost(guiCfg.Address()) && !guiCfg.InsecureSkipHostCheck {
-		// Verify source host
-		handler = localhostMiddleware(handler)
+	if !guiCfg.InsecureSkipHostCheck {
+		hasRemoteAddr := false
+		for _, addr := range guiCfg.Addresses() {
+			if !addressIsLocalhost(addr) {
+				hasRemoteAddr = true
+				break
+			}
+		}
+		// Verify source host if all listening addresses are local.
+		if !hasRemoteAddr {
+			handler = localhostMiddleware(handler)
+		}
```

Comment:
This should be a separate handler per listener, as now a single address taints all handlers.

Also, I am thinking that InsecureSkipHostCheck and UseTLS to be per address (which means we'd need Address to be a struct with additional parameters, not just a string)

Reply:
Fine